File: Adapters/MoPub/ObjC/Native/AvocarrotMoPubNativeCustomEvent.m
//
//  AvocarrotMoPubNativeCustomEvent.m
//  AvocarrotSDKAdaptersObjC
//
//  Created by Glispa GmbH on 14.11.17.
//  Copyright © 2017 Glispa GmbH. All rights reserved.
//

#import "AvocarrotMoPubNativeCustomEvent.h"
#import "AvocarrotMoPubCustomAdapter.h"
#import <AvocarrotCore/Avocarrot.h>
#import <AvocarrotNativeAssets/AvocarrotSDK+AVONativeAssets.h>

@interface AvocarrotMoPubNativeCustomEvent()

@property (nonatomic, strong) AvocarrotMoPubCustomAdapter *avocarrotCustomAdapter;

@end


@implementation AvocarrotMoPubNativeCustomEvent

- (void)requestAdWithCustomEventInfo:(NSDictionary *)info {
    if (!info) {
        NSLog(@"AvocarrotMoPubNativeCustomEvent - empty info!");
        [self requestDidFail];
        return;
    }
    
    NSString *adUnitId = info[@"adUnit"];
    if (!adUnitId) {
        NSLog(@"AvocarrotMoPubNativeCustomEvent - invalid adUnitId!");
        [self requestDidFail];
        return;
    }
    
    [AvocarrotSDK.sharedSDK loadNativeAdWithAdUnitId:adUnitId success:^UIView * _Nullable(AVONativeAssets * _Nonnull nativeAssests) {
        [[[nativeAssests onImpression:^{
            [self.avocarrotCustomAdapter registerImpressionToMopub];
        }] onClick:^{
            [self.avocarrotCustomAdapter registerClickToMopub];
        }] onLeftApplication:^{
            [self.avocarrotCustomAdapter registerFinishHandlingClickToMopub];
        }];
        
        self.avocarrotCustomAdapter = [[AvocarrotMoPubCustomAdapter alloc] initWithAd:nativeAssests];
        [self preloadImagesFromAd:nativeAssests];
        
        return nil;
    } failure:^(AVOError * _Nonnull error) {
        if (self.delegate && [self.delegate respondsToSelector:@selector(nativeCustomEvent:didFailToLoadAdWithError:)]) {
            [self.delegate nativeCustomEvent:self
                        didFailToLoadAdWithError:MPNativeAdNSErrorForInvalidAdServerResponse(error.localizedDescription)];
        }
    }];
}


- (void)preloadImagesFromAd:(AVONativeAssets *)ad {
    __block MPNativeAd *mopubAd = [[MPNativeAd alloc] initWithAdAdapter:(id <MPNativeAdAdapter>)self.avocarrotCustomAdapter];
    NSArray *images = [self getUrlsForImagesFromAd:ad];
    
    __weak AvocarrotMoPubNativeCustomEvent *weakSelf = self;
    
    [self precacheImagesWithURLs:images completionBlock:^(NSArray *errors) {
        if (errors) {
            if (weakSelf.delegate && [weakSelf.delegate respondsToSelector:@selector(nativeCustomEvent:didFailToLoadAdWithError:)]) {
                [weakSelf.delegate nativeCustomEvent:weakSelf
                            didFailToLoadAdWithError:MPNativeAdNSErrorForImageDownloadFailure()];
            }
        } else {
            if (weakSelf.delegate && [weakSelf.delegate respondsToSelector:@selector(nativeCustomEvent:didLoadAd:)]) {
                [weakSelf.delegate nativeCustomEvent:weakSelf didLoadAd:mopubAd];
            }
        }
    }];
}


- (void)requestDidFail {
    NSError *error = [[NSError alloc] initWithDomain:@"com.mopub.NativeCustomEvent" code:901 userInfo:nil];
    [self.delegate nativeCustomEvent:self didFailToLoadAdWithError:error];
}


- (NSArray<NSURL *> *)getUrlsForImagesFromAd:(AVONativeAssets *)ad {
    NSMutableArray<NSURL *> *urls = [NSMutableArray new];
    
    if (ad.iconURL || [@"" isEqualToString:ad.iconURL]) {
        NSURL *icon = [NSURL URLWithString:ad.iconURL];
        if (icon) {
            [urls addObject:icon];
        }
    }
    
    if (ad.imageURL || [@"" isEqualToString:ad.imageURL]) {
        NSURL *image = [NSURL URLWithString:ad.imageURL];
        if (image) {
            [urls addObject:image];
        }
    }
    
    return urls;
}

@end

File: Adapters/AvocarrotSDKAdapters/MoPub/AvocarrotMoPubNativeCustomEvent.swift
//
//  AvocarrotMoPubNativeCustomEvent.swift
//  AvocarrotSDKAdapters
//
//  Created by  Glispa GmbH on 15/06/16.
//  Copyright © 2016  Glispa GmbH. All rights reserved.
//

import Foundation
import MoPub
import AvocarrotNativeAssets

@objc(AvocarrotMoPubNativeCustomEvent)
open class AvocarrotMoPubNativeCustomEvent: MPNativeCustomEvent {

	fileprivate var avocarrotCustomAdapter: AvocarrotMoPubCustomAdapter?

	/** @name Requesting a Native Ad */

	/**
	 * Called when the MoPub SDK requires a new native ad.
	 *
	 * When the MoPub SDK receives a response indicating it should load a custom event, it will send
	 * this message to your custom event class. Your implementation should load a native ad from a
	 * third-party ad network.
	 *
	 * @param info A dictionary containing additional custom data associated with a given custom event
	 * request. This data is configurable on the MoPub website, and may be used to pass dynamic
	 * information, such as publisher IDs.
	 */
	open override func requestAd(withCustomEventInfo info: [AnyHashable: Any]!) {

		guard let adUnit = (info["adUnit" as NSObject] as? String) else { reportError("No AdUnit!"); return }

        AvocarrotSDK.shared.loadNativeAd(withAdUnitId: adUnit, success: {(nativeAd: AVONativeAssets) -> UIView? in

            nativeAd.onImpression({
                self.avocarrotCustomAdapter?.registerImpressionToMopub()
            })
            .onClick({
                self.avocarrotCustomAdapter?.registerClickToMopub()
            })
            .onLeftApplication({
                self.avocarrotCustomAdapter?.registerFinishHandlingClickToMopub()
            })

            self.avocarrotCustomAdapter = AvocarrotMoPubCustomAdapter(ad: nativeAd)
            self.preloadAdImages(forNativeAd: nativeAd)

            return nil
        }) {(_: AVOError) in
            self.reportError("AvocarrotSDK ad load error!")
        }
	}

	func reportError(_ msg: String) {
		delegate.nativeCustomEvent(self, didFailToLoadAdWithError: MPNativeAdNSErrorForInvalidAdServerResponse(msg))
	}

    /**
     Removes escape characters from the `urlString` and return an URL object.
     
     - Parameter urlString: string of the url
     - Returns: URL object
     */
    func escapeUrl(urlString: String) -> URL? {
        guard let escapedUrl = urlString.addingPercentEncoding(withAllowedCharacters: CharacterSet.urlQueryAllowed) else {
            return nil
        }

        return  URL(string: escapedUrl)
    }

    /**
     Transforms string urls to URL objects.
     
     - Parameter urlStrings: array of string urls
     - Returns: URL array object

    */
    func getImageUrls(urlStrings: [String?]) -> [URL] {
        var images = [URL]()
        for item in urlStrings {
            if let urlString = item {

                guard !urlString.isEmpty else {
                    continue
                }

                var url = URL(string: urlString)

                if url == nil {
                    url = escapeUrl(urlString: urlString)
                }

                guard let safeUrl = url  else {
                    continue
                }

                images.append(safeUrl)
            }

        }
        return images
    }

    func preloadAdImages(forNativeAd nativeAd: AVONativeAssets) {
        let mopubAd = MPNativeAd(adAdapter: avocarrotCustomAdapter)
        let images = getImageUrls(urlStrings: [nativeAd.iconURL, nativeAd.imageURL])

        precacheImages(withURLs: images, completionBlock: { [weak self] (error: [Any]?) in
            guard let sSelf = self else {return}
            if error != nil {
                sSelf.delegate.nativeCustomEvent(sSelf, didFailToLoadAdWithError: MPNativeAdNSErrorForImageDownloadFailure())
            } else {
                sSelf.delegate.nativeCustomEvent(sSelf, didLoad: mopubAd)
            }
        })
    }

}
